src/Language/Haskell/Liquid/Termination/Structural.hs

```diff
@@ -154,14 +160,59 @@ check env args = \case
     Coercion{} -> mempty
     Type{}     -> mempty
 
--- This is where we check a function call.
--- We go through the list of arguments as long as they are equal
--- to the original parameter. If we find one that is smaller, great. Otherwise,
--- fail.
-isGoodArgs :: [Param] -> [CoreArg] -> Bool
-isGoodArgs ps     (Cast e _ : as) = isGoodArgs ps (e : as)
-isGoodArgs ps     (Tick _ e : as) = isGoodArgs ps (e : as)
-isGoodArgs (p:ps) (Var v : as)
-    | v `elemVarSet` origParam p = isGoodArgs ps as
-    | v `elemVarSet` subterms p  = True -- yay!
-isGoodArgs _ _ = False
+data Call = Call
+  { callSource :: String
+  , candidateArgs :: S.HashSet Int
+  , decreasingArgs :: S.HashSet Int
+  } deriving (Eq, Show, Generic)
+
+instance Hashable Call
+
+-- This is where we  check a function call. We go through  the list of arguments
+-- and find the  indices of those which are decreasing.  Note that this approach
+-- is only guaranteed to  work when the arguments to the  function are named, so
+-- e.g.
+-- foo (x:xs) (y:ys) = foo xs (y:ys)
+-- won't necessarily work, but
+-- foo (x:xs) yys@(y:ys) = foo xs yys
+-- will.
+getStructArgs :: String -> [(Param, CoreArg)] -> Call
+getStructArgs src args =
+  let (ca, da) = getStructArgs' 0 args in Call src ca da
+  where
+    getStructArgs' _ [] = (S.empty, S.empty)
+    getStructArgs' i ((p, Cast e _) : args) = getStructArgs' i ((p, e) : args)
+    getStructArgs' i ((p, Tick _ e) : args) = getStructArgs' i ((p, e) : args)
+    getStructArgs' i ((p, Var v) : args)
+      | v `elemVarSet` origParam p = (S.singleton i, S.empty) <> getStructArgs' (i + 1) args
+      | v `elemVarSet` subterms p = (S.singleton i, S.singleton i) <> getStructArgs' (i + 1) args
+      | otherwise = getStructArgs' (i + 1) args
+    getStructArgs' i (_ : args) = getStructArgs' (i + 1) args
+
+-- Check if there is some way to lexicographically order the arguments so that
+-- they are structurally decreasing. Essentially, in order for there to be, we
+-- must be able to find some argument which is always either unchanged or
+-- decreasing. We can then remove every call where that argument is decreasing
+-- and recurse.
+structDecreasing :: Env () -> S.HashSet Call -> Result ()
+structDecreasing env calls
+  | S.null calls = mempty
+  | not $ S.null nonDecCalls = raiseErr nonDecCalls
+  | otherwise =
+      let ca = commonArgs calls in
+        if S.null ca then
+          raiseErr calls
+        else
+          structDecreasing env $ (S.fromList . mapMaybe (removeDecreasing ca) . S.toList) calls
+  where
+    raiseErr calls =
+      retErr env $ text "Non-structural recursion in calls" <+>
+      (text $ showPpr (S.map callSource calls))
+    nonDecCalls = S.filter (S.null . decreasingArgs) calls
+    commonArgs calls = foldl1 S.intersection (S.map candidateArgs calls)
```

Comment:
Code smell, sometimes you have `calls` as a parameter here, sometimes you just use the `calls` from `structDecreasing`. That’s confusing.

Reply:
This is a good point, I'll change it.